### DISCUSSION TITLE
Looking for simpler APIs

### DISCUSSION THREAD

**Discussion by wzzhu** [gomlx/gomlx#97] [Ideas]
It is a great project to combine go and XLA to build a pure go deep learning framework. It is the exact thing I have been looking for after using go. It is great that it could blur the CPU/GPU/TPU/etc hardware in the framework and deliver the bare metal performance.

After reading documents and code in the project and trying some examples on MacOS/arm platform, I would like to share some wishes for the APIs. My points may be wrong and biased. But I hope that bringing them out for discussion would make things better.

I use pseudo go code snippet to express what I wish to write a trainer using gomlx. 

```

// err handling omitted

model, err := gomlx.NewModel(device)  // backend, device, graph are all managed internally by the model

shape := []int{10, 5}
rng := rand.NewRand(seed)
weights := model.NewTensor("weights", dtype, shape, gomlx.NewRandomDataInitializer(rng))
bias := model.NewTensor("bias", dtype, []int{5}, gomlx.NewRandomDataInitializer(rng))

x := model.NewPlaceholder("x", dtype, []int[100, 10])
y := model.NewPlaceholder("y", dtype, []float64{5})
output := model.Mul(weights, x).Add(bias) // Last func may also be called BroadcastAdd

loss := model.MeanSquaredError(output, y)
optimizer := gomlx.NewAdamOptimizer(learningRate, ...)

inputs:= []float64 {....} // one-dimensional flattened data
labels := []float64{ ....}
x.Assign(inputs)
y.Assign(labels)
results, gradients, err := model.Eval([]gomlx.Tensor{loss})
optimizer.Backward(gradients)
```

// Load and Save, including both meta and weights
model.Save(filename)
model2, err := gomlx.LoadModel(filename)
...

A few points from above example:

- Less packages to import and less APIs to make it simple and more attractive. It hides the implantation details of backend/computation graph/etc from user, and uses a single context aware model (can use gomlx.Context too) to track all computation flows instead of passing graph, ctx, backends back and forth in the calls. 
- Simple and clean primitive APIs to allow user to access the gradients and write the explicit training loop (can provide some simple trainer that provides the normal training loop). 
- High level API like keras.Layers built upon the primitive APIs that allow faster construction of neural network model (Not in the example, but could be imagined).
- Using a simple tensor data management by using a flattened float slice to supply data to tensors.
- Allow model to load/save without needing to build the graph (saw discussion on other threads).

Thanks!

**Comment by janpfeifer** [comment]
Thanks for checking it out @wzzhu . 

Funny, for the most part your example is not very different than the current set up, if you "squint" (== getting used with GoMLX). Consider the following:

- What you call `Model`, GoMLX calls `graph.Graph`: a Model is a computation graph. But not all computation graphs are ML models. One can use it to accelerate some physics simulation, for instance. So just find&replace "model" by "graph".
- You don't need to prefix everything with model, like in `model.Mul(x, weights)`. Since `x` and `weigths` objects (in GoMLX they are called `Node`) have a link to the `Graph` (model), you can simply do `Mul(x, weights)`, it knows which graph it's referring to.
- Ideally, you want to separate the `loss` from the "model", at least in ML. Because the loss is only used in training, not in inference.
- To support training loops, one needs to define a concept of a "Dataset" to train, and its interface must be tied to the definition of what a ML models is (now the more specific "model", not a more generic "graph"). GoMLX uses [`train.ModelFn`](https://pkg.go.dev/github.com/gomlx/gomlx@v0.15.3/ml/train#ModelFn) as a model -- a model is just a function that creates the model, given inputs, and provides outputs.
- Variables of the models need lots of flexibility: batch normalization, freezing parts of the model, optimizers variables (moving averages) require the ability to introspect on the variables, separate "training" from "inference" (the models are actually different in both cases). This is really not trivial if one wants to keep things sane and flexible.

For a long time I had to use Keras, which makes things look pretty, but it was impossible (or super hacky) to do anything that was different than the standard demo models, it was a major pain. 

One of core ideas of GoMLX is to allow flexibility, composability (for instance the training loop), and transparency. At the cost of making it more complex at times (more parts), and more verbose. 

So it favors those in the real world creating models (or fine-tuning) with specific non-standard needs, at the detriment of the casual student that only wants to implement the trivial examples and make it look "clean" (quotes because it's a very ambiguous term). I hope it can fulfill both needs, but the former takes precedence.

About your points:

> * Less packages to import and less APIs to make it simple and more attractive. 

Yes, there are many packages 😞 But I think this is a Go thing. GoMLX is modular, and the price to pay is to have to import all parts needed. Fortunately, modern IDEs will do it automatically.

> It hides the implantation details of backend/computation graph/etc from user, and uses a single context aware model (can use gomlx.Context too).

That is not that easy. Because in some use cases one doesn't want to use a Context -- again for non-ML use cases. It's the price to pay for composability.

I'm not sure I follow "hide the implementation details of backend/computation graph"  ? Maybe I'm missing something, could you provide an example ?

>  to track all computation flows instead of passing graph, ctx, backends back and forth in the calls.

Could you provide examples ? I thought it was passing these only exactly where needed, but maybe I'm missing something. 

Explicit examples are always helpful, because they are actionable, it allows me to make fixes, where possible 😄 

> * Simple and clean primitive APIs to allow user to access the gradients and write the explicit training loop (can provide some simple trainer that provides the normal training loop).

And I thought the primitives in [`train`](https://pkg.go.dev/github.com/gomlx/gomlx@v0.15.3/ml/train) were clean (and relatively well documented), albeit not simple.

But back to the point of Keras: I don't want to sacrifice flexibility just to make the standard example look pretty.

Again, explicit examples here would be helpful to make it actionable.

> * High level API like keras.Layers built upon the primitive APIs that allow faster construction of neural network model (Not in the example, but could be imagined).

Ha, Keras layers is full of issues, if you just need to dig a bit deeper 😄 ... I did them for years of my life.

Now GoMLX does provide high level layers (see `fnn`, `kan`, `lstm`, `layers.Convolution`, `layers.MultiHeadAttention`). Definitely, there could be more, and contributions are welcome.

But think: what do you really want/need from Keras layers ? Can you give concrete examples ?

> * Using a simple tensor data management by using a flattened float slice to supply data to tensors.

Please, check out the package [`tensors`](https://pkg.go.dev/github.com/gomlx/gomlx@v0.15.3/types/tensors).

> * Allow model to load/save without needing to build the graph (saw discussion on other threads).

So this is definitely desirable in some use cases. I don't know yet what format to actually use -- maybe ONNX (likely), maybe StableHLO (or the dreaded Tensorflow SavedModel format). And this is another case that Keras introduced  major pains in my previous work. It's an issue in search of a good solution still.

Btw, do you have an immediate need right now, that cannot be solved by using the model checkpoint with a version of your model Go code ?

Also, so we can have a more in-depth discussion, I highly recommend reading careful through one of the full end-to-end examples. For instance, take the [UCI Adult Census example](https://github.com/gomlx/gomlx/blob/main/examples/adult/demo/main.go). The model is mostly two lines, all the rest is data and feature pre-processing (embeddings, normalizations) -- as is almost always the case in ML, 95% of the code is data manipulation: 

```
logits := Concatenate(allEmbeddings, -1)
logits = fnn.New(ctx.In("fnn"), logits, 1).Done()
````

**Reply by wzzhu** [reply]
Thanks for the prompt and long reply which includes many details!
My discussion below would be casual (and would make many mistakes :)

In my daily work I train and deploy neural network models for the production use of quantitive trading. As our code base is almost in go, I always keep an eye on emerging deep learning frameworks in go. Ideally I wish to find a deep learning framework in go that can do either:
1. A minimal and stable framework that can do high performance AutoGrad.
2. A full fledged framework that can be used like Tensorflow/PyTorch/JAX/etc.

I tried gorgonia which was the only framework previously to support both training and serving in go.  I read most of its source code and understand how it does autograd and other parts. I occasionally contributed to some bug fixes. Internally in my own work, we wrapped gorgonia with simple APIs like what I wrote in the example snippets above and keras-like layer abstraction below

model, err := nn.NewModel(nn.Param{
		Spec: &nn.NetworkSpec{
			LayerParams: []nn.LayerParam{
				nn.InputLayerParam{Shape: 28 * 28},
				nn.DenseLayerParam{OutputShape: 128, Activation: "relu", Dropout: 0.2},
				nn.DenseLayerParam{OutputShape: 10, Activation: "softmax"},
			}}},...)

 
For advanced network structure tweaking, I turn to tensorflow and write python code, using SavedModel format to load from go code. Tensorflow has easier and ready to use building blocks for advanced network structures like CNN, RNN, LSTM, GRU, Attention network, etc. And sometimes I can use customized computation and loss using python class derived from tf.keras.Model

The project gomlx caught my eye due to its integration with XLA and the active and passionate contributors 
(I found installation on Mac M1/2 failed before, but a few days later when I came back, it was fixed and works well!) 

> * What you call `Model`, GoMLX calls `graph.Graph`: a Model is a computation graph. But not all computation graphs are ML models. One can use it to accelerate some physics simulation, for instance. So just find&replace "model" by "graph".
> * You don't need to prefix everything with model, like in `model.Mul(x, weights)`. Since `x` and `weigths` objects (in GoMLX they are called `Node`) have a link to the `Graph` (model), you can simply do `Mul(x, weights)`, it knows which graph it's referring to.
> * Ideally, you want to separate the `loss` from the "model", at least in ML. Because the loss is only used in training, not in inference.

In my "Model", I mean the manager of the computation graph, backend engine for execution, and the model meta data. So it is a monolithic thing. It can be used both for evaluation and training (if setting something like IsTraining flag to true). The graph.Graph is a general purpose computation graph and model would just focus on using it for neural network.

Yes, loss can be separated from model ideals. Or it can be inside the model as it is one node in the graph inside the model.

> * To support training loops, one needs to define a concept of a "Dataset" to train, and its interface must be tied to the definition of what a ML models is (now the more specific "model", not a more generic "graph"). GoMLX uses [`train.ModelFn`](https://pkg.go.dev/github.com/gomlx/gomlx@v0.15.3/ml/train#ModelFn) as a model -- a model is just a function that creates the model, given inputs, and provides outputs.
> * Variables of the models need lots of flexibility: batch normalization, freezing parts of the model, optimizers variables (moving averages) require the ability to introspect on the variables, separate "training" from "inference" (the models are actually different in both cases). This is really not trivial if one wants to keep things sane and flexible.
> 
> For a long time I had to use Keras, which makes things look pretty, but it was impossible (or super hacky) to do anything that was different than the standard demo models, it was a major pain.
> 
> One of core ideas of GoMLX is to allow flexibility, composability (for instance the training loop), and transparency. At the cost of making it more complex at times (more parts), and more verbose.
> 
> So it favors those in the real world creating models (or fine-tuning) with specific non-standard needs, at the detriment of the casual student that only wants to implement the trivial examples and make it look "clean" (quotes because it's a very ambiguous term). I hope it can fulfill both needs, but the former takes precedence.

Ha, I hope that it can serve a wider audience, since most of the time, the standard building blocks without deep hacking can already have a lot of practical uses in real world.

> 
> About your points:
> 
> > * Less packages to import and less APIs to make it simple and more attractive.
> 
> Yes, there are many packages 😞 But I think this is a Go thing. GoMLX is modular, and the price to pay is to have to import all parts needed. Fortunately, modern IDEs will do it automatically.

I still think it is overkill and creates many shallow sub packages like what Java does :) It is so overwhelming of gomlx packages in the user go file. While IDE can auto import, human mind cannot remember so many package names when coding. Current settings may be good for reader (with the help of IDE), but not for writer (who need to come up with the right package name first). I guess current arrangement of sub packages will scare many users away of trying it.

> 
> > It hides the implantation details of backend/computation graph/etc from user, and uses a single context aware model (can use gomlx.Context too).
> 
> That is not that easy. Because in some use cases one doesn't want to use a Context -- again for non-ML use cases. It's the price to pay for composability.
> 
> I'm not sure I follow "hide the implementation details of backend/computation graph" ? Maybe I'm missing something, could you provide an example ?

For backend hiding, I mean that the programmer doesn't need to care how it is run on the backend (thanks to XLA). Programmers just need to specify the device ID (like cpu:0, gpu:0) without worrying the memory management or scheduling of the computation on physical devices.

So the lib would just let user to mention about backend to use CPU, for example, then user don't need to remember the backend any more in the rest of the code.

> 
> > to track all computation flows instead of passing graph, ctx, backends back and forth in the calls.
> 
> Could you provide examples ? I thought it was passing these only exactly where needed, but maybe I'm missing something.
> 
> Explicit examples are always helpful, because they are actionable, it allows me to make fixes, where possible 😄
> 

For example, in https://github.com/gomlx/gomlx/blob/main/examples/linear/linear.go, it keeps mentioning backends in many function calls.

```
	backend := backends.New()

	trueCoefficients, trueBias := initCoefficients(backend, *flagNumFeatures)
	fmt.Printf("Target coefficients: %0.5v\n", trueCoefficients.Value())
	fmt.Printf("Target bias: %0.5v\n\n", trueBias.Value())

	inputs, labels := buildExamples(backend, trueCoefficients, trueBias, *flagNumExamples, *flagNoise)
	fmt.Printf("Training data (inputs, labels): (%s, %s)\n\n", inputs.Shape(), labels.Shape())

	// Create an in-memory dataset from the tensors.
	dataset := must.M1(data.InMemoryFromData(backend, "linear dataset", []any{inputs}, []any{labels})).
		Infinite(true).Shuffle().BatchSize(*flagNumExamples, false)
	// dataset := &Dataset{"training", []*tensors.Tensor{inputs}, []*tensors.Tensor{labels}}

```


> > * Simple and clean primitive APIs to allow user to access the gradients and write the explicit training loop (can provide some simple trainer that provides the normal training loop).
> 
> And I thought the primitives in [`train`](https://pkg.go.dev/github.com/gomlx/gomlx@v0.15.3/ml/train) were clean (and relatively well documented), albeit not simple.
> 
> But back to the point of Keras: I don't want to sacrifice flexibility just to make the standard example look pretty.
> 
> Again, explicit examples here would be helpful to make it actionable.
> 
For trainer loop, I need explicit loop when I implement reinforcement learning algorithms, e.g., policy gradient or others, it needs to customize the gradients and play with rewards. And for some other cases, it may need to clip the gradients (I haven't checked if gomlx supports clipping in parameters).

> > * High level API like keras.Layers built upon the primitive APIs that allow faster construction of neural network model (Not in the example, but could be imagined).
> 
> Ha, Keras layers is full of issues, if you just need to dig a bit deeper 😄 ... I did them for years of my life.
> 
> Now GoMLX does provide high level layers (see `fnn`, `kan`, `lstm`, `layers.Convolution`, `layers.MultiHeadAttention`). Definitely, there could be more, and contributions are welcome.
> 
> But think: what do you really want/need from Keras layers ? Can you give concrete examples ?

Keras layer abstraction are not that critical for my needs. Normally I use keras class style API to derived model class from tf.keras.Model, it can already support very flexible customized computation. I use tf.keras.layers when constructing standard network.

> 
> > * Using a simple tensor data management by using a flattened float slice to supply data to tensors.
> 
> Please, check out the package [`tensors`](https://pkg.go.dev/github.com/gomlx/gomlx@v0.15.3/types/tensors).
> 
> > * Allow model to load/save without needing to build the graph (saw discussion on other threads).
> 
> So this is definitely desirable in some use cases. I don't know yet what format to actually use -- maybe ONNX (likely), maybe StableHLO (or the dreaded Tensorflow SavedModel format). And this is another case that Keras introduced major pains in my previous work. It's an issue in search of a good solution still.
> 
> Btw, do you have an immediate need right now, that cannot be solved by using the model checkpoint with a version of your model Go code ?

Not that urgent. Since if I use glmlx more deeply, I can wrap with own format. In my discussion above, I wrap gorgonia's model using gob format. So I don't use ONNX (I did try converting some PyTorch models to it and convert it back to tf.SavedModel which I can use using go).

> 
> Also, so we can have a more in-depth discussion, I highly recommend reading careful through one of the full end-to-end examples. For instance, take the [UCI Adult Census example](https://github.com/gomlx/gomlx/blob/main/examples/adult/demo/main.go). The model is mostly two lines, all the rest is data and feature pre-processing (embeddings, normalizations) -- as is almost always the case in ML, 95% of the code is data manipulation:

Yes, thanks! I try to step it through debugger so that I can have a more clear understanding of gomlx.

Lastly, I still fell that too many imports hurt :) but it is not critical as user can wrap it in one package .
Last wish: Simplify everything whenever possible :)
Please keep the great work!
> 
> ```
> logits := Concatenate(allEmbeddings, -1)
> logits = fnn.New(ctx.In("fnn"), logits, 1).Done()
> ```

**Reply by janpfeifer** [reply]
Thanks for the feedback @wzzhu . I hear you, but at a high level/generically it is hard to be actionable. Still it's interesting to talk about to guide future choices.

I few comments below:

The example you gave:

```go
model, err := nn.NewModel(nn.Param{
Spec: &nn.NetworkSpec{
LayerParams: []nn.LayerParam{
nn.InputLayerParam{Shape: 28 * 28},
nn.DenseLayerParam{OutputShape: 128, Activation: "relu", Dropout: 0.2},
nn.DenseLayerParam{OutputShape: 10, Activation: "softmax"},
}}},...)
```

GoMLX is less object-oriented, and tend to be more functional. The equivalent would be something like:

```go
func Model(ctx *context.Context, spec any, inputs []*Node) []*Node {
   batchSize := inputs[0].Shape().Dim(0)
   input := Reshape(inputs[0], batchSize, -1)  // Flatten image.
   logits := fnn.New(ctx.In("fnn"), input, 10).
      NumHiddenLayers(1, 128).
      Activation(activations.TypeRelu).
      Done()
   probabilities := Softmax(logits)
   return []*Node{probabilities}
}
```

It's very similar, except it's functional as opposed to an configuration object (the first version "model" object) that later is converted to the actual computation graph. In a way, the functional form has one less layer of indirection.


---
## Passing `backend` everywhere

Indeed the linear example particularly use `backend` a lot: it needs it for 4 things:

- To generate the coefficients to be learned.
- To generate random data to train with some added noise.
- The "InMemoryDataset" does operations with GoMLX (splitting batches, etc) so it needs to know what is the backend.
- Training needs the backend to train.

Since these four tasks are using GoMLX for numeric operations, hence they need a backend. It's one of those concepts that is trivial, but it comes with the framework.

One option to hide is using a global: I'm against it because sometimes there may be more than one backend in the same system (multi-gpu, gpu/cpu running in parallel), and because I avoid as much as I can hidden parameter. Generally, the more functional, the easier to make static analysis of the code, as one is reading it.

The other option is to package many arguments that are often used together in one container. I do this quite a bit in some cases. But I didn't see in `linear.go` where I could do it. 

Any other suggestions ?

---
## Functionality, Gradient Clipping and Reinforcement Learning 

In terms of functionality, GoMLX now supports CNNs (Convolutions), LSTMs, Attention (MultiHeadAttention), GNNs, KANs (4 different versions), Deep Retrieval, Two-Tower models (or any frankenstein model), multi-tasking, fine-tune LLMs, Stable Diffusion,  FFT (with auto-diff) for learning on the frequency domain, Batch and Layer normalization, learnable PieceWise-Linear, BSpline and Rational Functions, etc. Definitely not PyTorch or Jax/TensorFlow, but a lot is there. And the basis is there to implement most things missing (GRU would be trivial to implement).

> For trainer loop, I need explicit loop when I implement reinforcement learning algorithms, e.g., policy gradient or others, it needs to customize the gradients and play with rewards. And for some other cases, it may need to clip the gradients (I haven't checked if gomlx supports clipping in parameters).

It does support clipping gradients or arbitrary functions that update variables to be run after optimization. For instance, I use this when I want to project a function (bspline or piecewise-linear) to monotonicity after a gradient is applied. This way I can guarantee the model learned is monotonic with respect to some input, for instance. 

I use it in PiecewiseLinear-KAN models with learnable split points, to not allow split points to cross each other -- so I project them to monotonicity + a margin. See code for projection here: 

* https://github.com/gomlx/gomlx/blob/main/ml/train/optimizers/optimizers.go#L311

And how it schedules this to be executed after the updates here (a few lines):

* https://github.com/gomlx/gomlx/blob/main/ml/layers/kan/piecewiselinear.go#L190

Loops in Reinforcement Learning are not straight forward -- from having implemented an AlphaZero for the Hive game a long time ago. But GoMLX is very flexible. The `train.Train` object helps create a `TrainStep` method that calls optimizers, regularizers, post-optimizer registered functions. The `train.Loop` simply loops over `TrainStep` in various ways, and support callbacks for arbitrary things like a progress-bar, plotting of loss during training, or periodic checkpoint saving. They are very configurable, flexible, and can be replaced or not used if one doesn't want to.

--- 
## Fine-grained splitting of packages

This is full of gray areas (different people will weigh the pros/cons differently) ... I don't believe there is an absolute right/wrong. But let me put forward my own arguments/factors I used when deciding to split into separate packages, only so you understand, even if you disagree:

- I use GoLand, and I never type any import. I type the package name, the IDE automatically adds the import line, and the auto-complete immediately picks it up. So when I type `activations.`, the IDE immediately imports the `activations` package and it already offer auto-complete with the various activation functions (Sigmoid, Relu, Selu, Gelu, Swish, LeakyRelu, ...). I agree, without a good IDE, this would be a pain ... Having said that, I never code a model from scratch, I'm always copy&pasting from my previous model, where all things are already imported, and I just change code.
- Packages provide the equivalent of "namespaces" in C++.  It makes it easier to find code and functions related to areas of need. If I need a metric, I go to the `metrics` package, I don't need to search around for metrics in a giant package with everything. I type `metrics.` and the IDE will immediately show me all the metrics available ...
- Large packages instead require prefixing (or suffixing) stuff with the same strings. For instance: if `Context` were in the same package as `graph` (where all ops are), creating a new Context would need to be called `NewContext`. But since it is in its own package, it becomes `context.New()`. Reading it's the same, just the order of the works "context" changed. So if everything in the package starts becoming prefixed / suffixed to their own concerns ... it's a point to consider splitting.
- Separate usages: when functionality is large, and not everyone needs it, I don't want to force one to link (or care about) the unused stuff. E.g.: the `graph` package is all one needs to build an accelerated graph, for instance to do physics simulation, or some number crunching task. They don't need to know anything about contexts, variables and hyperparameters (all these are in the `context` package). So different primary use-cases helped me guide on splits as well.

If you start using and see a good case where packages should be merged, pls let me know. There is a cost of changing, but things are not set in stone yet -- plus I'm sure I did many bad choices, I just don't know which are those 😄 

---
> Last wish: Simplify everything whenever possible :)

I quite agree here ... but "simplify" is very ambigous. Go is considered a very simple language, but it's much more verbose than Python and Javascript.

I like the simplicity of Go: my understanding is that it focus on making code easy to understand, with no surprises, even if more verbose at time.

With GoMLX I try to do the same: the point of having more parameters at times, I think it's a simplification. One knows what is happening, there is no magic so to say -- which generally leads to trouble later, or misunderstood concepts.

I mean that is the aspiration ... I took my own share of shortcuts in some places 😳

**Comment by janpfeifer** [comment]
Let me add a bit to the answer, from a high level:

GoMLX doesn't hide things (Backend, Graph, Context) deliberately, because I believe they are important. 

It does add complexity: they are not particularly hard concepts, but they are more concepts, that need to be learned, so more investment up-front from the user. 

But after years (actually decades) of doing NNs, hiding these things behind magic, like Keras does, often comes with the terrible cost  of surprising behavior (lots of ["WAT"](https://www.destroyallsoftware.com/talks/wat) -- we had a running document of WAT for Keras), that is hard to work around and requires hacks that break over versions, etc. And who pays for this price are those doing most of the ML work and research. 

The casual student/curious programmer that only wants to do a very standard model won't notice it, so the "clean" (relative term) API looks awesome for them.
